### DISCUSSION TITLE
Add query parameter to all data requests

### DISCUSSION THREAD

**Discussion by liegeandlief** [remix-run/remix#2223] [Q&A] [answered]
I would like to add a query parameter to all data requests made from the client (i.e. all loader and action requests). This is so that I can append a build number to the requests so that I know that cached responses won't be used when a new build is deployed.

Is there any way of interceping data requests in order to add this parameter?

**Comment by kiliman** [comment, marked as answer]
I created a codesandbox that shows how you can replace the existing `fetch` function with a custom function that adds the search param. You can pass the build id in your root loader and place on the `window` object.

```ts
import { hydrate } from 'react-dom'
import { RemixBrowser } from 'remix'

if (!window.__fetch) {
  window.__fetch = window.fetch
  window.fetch = (url, options) => {
    let request
    let newUrl
    if (url instanceof Request) {
      request = url
      newUrl = request.url
    } else {
      newUrl = typeof url === 'string' ? new URL(url) : url
    }
    if (newUrl.searchParams.has('_data')) {
      // remix data request
      newUrl.searchParams.set('__build', window.ENV.REMIX_BUILD_ID)
    }
    return window.__fetch(request ?? newUrl, options)
  }
}

hydrate(<RemixBrowser />, document)
```
https://codesandbox.io/s/replace-fetch-og0myd?file=/app/entry.client.tsx

**Reply by liegeandlief** [reply]
Thank you, that's really helpful! :)
